### ISSUE TITLE
Allow explicit call ordering within a pallet

### ISSUE THREAD

**Issue by joepetrowski** [paritytech/substrate#10408]
Similar to explicit ordering in the `construct_runtime` macro, we could annotate calls within a pallet with an index. This would have a few advantages, namely the ability to remove a function without affecting the indices of other calls.

This could be useful in pallet versioning and transaction version, where a pallet's version would need to increment when there is a breaking change to these indices/function signatures. By leaving call indices empty on function removal, wallets/signers don't need to worry about signing an unintended call as it would be null.

Transaction version could either increment on pallet version changes or be reserved for changes in the serialization (e.g. signed extensions).

**Comment by apopiak** [comment]
IMO the `transaction_version` would still need to change on removal, but the migration would be trivial.

**Comment by bkchr** [comment]
> IMO the `transaction_version` would still need to change on removal, but the migration would be trivial.

No, because there would be no harm. The call would fail to decode, aka nothing serious would happen.

**Comment by KiChjang** [comment]
Was trying to implement this lately, but found out that [enum discriminants on non-unit variants are experimental](https://github.com/rust-lang/rust/issues/60553). We'll need to figure out another mechanism for associating an index with every call in the enum, and use that to encode/decode correctly.

**Comment by xlc** [comment]
scale codec does support macro attributes to specific encode index

**Comment by bkchr** [comment]
Yeah and we use exactly this in `construct_runtime!` for pallets.
